src/uct/tcp/tcp_sockcm_ep.c

```diff
@@ -357,6 +352,21 @@ ucs_status_t uct_tcp_sockcm_ep_server_handle_data_received(uct_tcp_sockcm_ep_t *
 
         /* server to wait for any notification (disconnect) from the client */
         events = UCS_EVENT_SET_EVREAD;
+    } else if (ucs_test_all_flags(cep->state, UCT_TCP_SOCKCM_EP_ON_SERVER |
+                                              UCT_TCP_SOCKCM_EP_DATA_RECEIVED) &&
+               (!(cep->state & UCT_TCP_SOCKCM_EP_SERVER_CREATED))) {
+        status = uct_tcp_sockcm_ep_server_invoke_conn_req_cb(cep);
+        if (status != UCS_OK) {
+            goto out;
+        }
+
+        /* server to listen to disconnect events and send its private data to
+         * the client only if its ep was already created */
+        if (cep->state & UCT_TCP_SOCKCM_EP_SERVER_CREATED) {
+            events = UCS_EVENT_SET_EVWRITE | UCS_EVENT_SET_EVREAD;
+        } else {
+            events = UCS_EVENT_SET_EVREAD;
```

Comment:
seems like we can check this outside the 'if', right before ucs_async_modify_handler(), and also remove line 354:
events = UCS_EVENT_SET_EVREAD;
if ((cep->state & UCT_TCP_SOCKCM_EP_SERVER_CREATED) && 
   !(cep->state &UCT_TCP_SOCKCM_EP_CONN_SENT)) { 
   events |= UCS_EVENT_SET_EVWRITE
}